src/net/conn.c

```diff
@@ -0,0 +1,122 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <postgres.h>
+#include <pg_config.h>
+
+#include "conn_internal.h"
+
+static ConnOps *conn_ops[_CONNECTION_MAX] = {NULL};
+
+static const char *conn_names[] = {
+	[CONNECTION_PLAIN] = "PLAIN",
+	[CONNECTION_SSL] = "SSL",
+	[CONNECTION_MOCK] = "MOCK",
+};
+
+static Connection *
+connection_internal_create(ConnectionType type, ConnOps *ops)
+{
+	Connection *conn = palloc(ops->size);
+
+	if (NULL == conn)
+		return NULL;
+
+	memset(conn, 0, ops->size);
+	conn->ops = ops;
+	conn->type = type;
+
+	return conn;
+}
+
+Connection *
+connection_create(ConnectionType type)
+{
+	Connection *conn;
+
+	if (type == _CONNECTION_MAX)
+		elog(ERROR, "ts::net: invalid connection type");
+
+	if (NULL == conn_ops[type])
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("ts::net: %s connections are not supported", conn_names[type]),
+				 errhint("Enable %s support when compiling the extension.", conn_names[type])));
+
+	conn = connection_internal_create(type, conn_ops[type]);
+
+	Assert(NULL != conn);
+
+	if (NULL != conn->ops->init)
+		if (conn->ops->init(conn) < 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_CONNECTION_EXCEPTION),
```

Comment:
I don't think we should use the connection exception here as I think it normally means a connection to the DB, and this seems to be stretching that a bit far? I think just a normal internal error makes more sense...